### PR TITLE
Use GitHub Actions for CI

### PR DESCRIPTION
**Pull request by coldlink**
## What does this change?

Switches CI from TeamCity to GitHub Actions (with the use of https://github.com/guardian/actions-assume-aws-role).

There are a few benefits to using GHA individually:
- No build queue
- No need to manage `node` docker versions
- No need to globally install `yarn`
- Build logs are co-located with the PR (via the Checks tab)
- Collectively, use of GHA means we don't have to maintain our own CI infrastructure.

In RiffRaff the build numbers will count up from 5000 now, as the teamcity builds are currently in the 4000s.

To help perform the migration, I used [migration doc](https://docs.google.com/document/d/1Xz-gzptICIza0oio0Kza3LWg2a6n_015o0MPu0LGfuU/view#) and example prs from [Prism](https://github.com/guardian/prism/pull/259) and [tag-janitor](https://github.com/guardian/tag-janitor/pull/213).

## How to test

`identity:identity-gateway` should appear as a deployable project in Riff-Raff. It should be successfully deployed to CODE.

https://riffraff.gutools.co.uk/deployment/view/efa79594-2d2f-4b0d-8339-3e591bf3873f

## Pre merge
- Change the branch protection rules to remove TeamCity from required, and add the CI Github Action to required

## Post merge
- Ask people to rebase their branches on `main` to deploy to `CODE`/`PROD` going forward
- Remove project from TeamCity

### CODE REVIEW

File: ci.sh [hunk 1]
```diff
@@ -2,7 +2,18 @@
 
 set -ae
 
+if [[ ! -z "${TEAMCITY_VERSION}" ]]; then
+  echo "Running in TeamCity. Nope!"
```
Comment (ob6160) [inline ci.sh:6]: Haha brutal 😆

**Review by ob6160 — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      id-token: write
+      contents: read
+    steps:
+      - uses: actions/checkout@v2
+      - uses: guardian/actions-assume-aws-role@v1.0.0
+        with:
+          awsRoleToAssume: ${{ secrets.GU_RIFF_RAFF_ROLE_ARN }}
+      - uses: actions/setup-node@v2
+        with:
+          node-version: '16'
```
Comment (akash1810) [inline .github/workflows/ci.yml:18]: https://github.com/actions/setup-node/pull/338 allows you to specify a `node-version-file`, which makes it easier to keep things inline.

File: ci.sh [hunk 3]
```diff
@@ -2,7 +2,18 @@
...
+  exit 0
+fi
```
Comment (akash1810) [inline ci.sh:8]: Not needed, but is a good marker to see which PR you were inspired by 😄 .

File: .github/workflows/ci.yml (hunk 2 above)
Comment (coldlink) [reply .github/workflows/ci.yml:18]: Thanks! Even better 😃

File: ci.sh (hunk 3 above)
Comment (coldlink) [reply ci.sh]: Haha, I'll remove it.

File: ci.sh [hunk 4]
```diff
@@ -2,7 +2,13 @@
...
 
+# Don't remove this line as it's used to number the build based on the
+# last build number in teamcity
+LAST_TEAMCITY_BUILD=5000
+export GITHUB_RUN_NUMBER=$(( $GITHUB_RUN_NUMBER + $LAST_TEAMCITY_BUILD ))
+
 yarn install --frozen-lockfile
 yarn lint
 yarn test
 yarn build
```
Comment (dominickendrick) [inline ci.sh:13]: Do we also want to add `yarn tsc` in here, as its been removed from the other action ...

Comment (coldlink) [reply ci.sh:13]: `yarn test` actually already runs `yarn tsc`

https://github.com/guardian/gateway/blob/85ec804a51aa1b5a8c669729d7d8702bccc53079/package.json#L18

Comment (coldlink) [reply ci.sh:13]: we could separate it out into two separate commands for clarity?

Comment (coldlink) [reply ci.sh:13]: or combine the linting into the test command too as it's a kind of test?

```
"test": "yarn lint && yarn tsc && yarn test:unit", 
```

Comment (dominickendrick) [reply ci.sh:13]: Hmm, If they are all seperate commands, then when build fails, you may get more clarity about the step that is an issue, whereas if they are all combined, its harder to work out what failed ?

Comment (coldlink) [reply ci.sh:13]: internally yarn just pipes those commands to the shell, so it will still show on which command it failed on

Comment (dominickendrick) [reply ci.sh:13]: ok, so combining them seems like the right approach : )

Comment (coldlink) [reply ci.sh:13]: Code_LrZ5LxP8EA

**Review by dominickendrick — APPROVED** [review]

**Pull request merged**